### ISSUE TITLE
How to use ember-validations without object controllers?

### ISSUE THREAD

**Issue by arenoir** [DockYard/ember-validations#294]
I am trying to move away from object and array controllers and and having trouble finding documentation on how best to implement ember-validations in a world without object controllers.  

I don't think validations belong in the model or the component.  I am leaning toward using a ObjectProxy and keeping track of it using stefanpenners [state-services](https://github.com/stefanpenner/ember-state-services) addon.

**Comment by danmcclain** [comment]
We're using it with an `Ember.Controller` so that when routable components land, it should be fairly trivial for us to swap the controllers for components. I can put together an example tomorrow morning of how we are wiring this up

**Comment by arenoir** [comment]
@danmcclain I think using a controller and eventually a routable component is a valid solution.  However I think it is better to be able to maintain the validation state and changes of model attributes between components and routes.

I am trying to wrap my head around pushing objects to a service in some sort of accountable way.

**Comment by jackmatt2** [comment]
@danmcclain would you not also need to explicitly use `model.` in the validations hash?

```javascript
validations : {
    model.firstName: {
      presence: true,
    },
}
```

given the model properties are no longer proxied to the controller in `Ember.Controller`.

**Comment by danmcclain** [comment]
@jackmatt2 Correct

**Comment by danmcclain** [comment]
@arenoir https://gist.github.com/danmcclain/78b5182de8493b21b462

**Comment by jreading** [comment]
We are experiencing the same issue and the solution suggested in the gist doesn't not seem to work for us.

**Comment by bcardarella** [comment]
@jreading can you describe what isn't working?

**Comment by jreading** [comment]
tl;dr: False alarm.

These may "work" after all, just seems like unexpected behavior.

Validations that worked with ObjectController:

```
validations: {
        name: {
            presence: {
                "if": "hasName"
            }
        }
}
```
No longer work.

Trying the example in the gist above by aliasing the model in the controller:
```
validationModel: Ember.computed.alias("model"),

...

validations: {
        "validationModel.name": {
            presence: {
                "if": "hasName"
            }
        }
}
```
Puts the error in:

```
errors.validationModel.name
```
My bad, all our the observers for errors.name that we have weren't firing for obvious reasons, so just need to refactor those or alias validationModel.errors.
:neutral_face:

**Comment by jackmatt2** [comment]
mmmm, I can't seem to get this working either.

```javascript
export default Ember.Controller.extend(EmberValidations.Mixin, {

  validationModel: Ember.computed.alias("model"),

  validations : {
    'validationModel.bookingSource.content' : {
      presence: true
    },
  }
});
```

bookingSource is an async belongsTo ember-data model relationship.

I am getting the followign error

```
Error while processing route: booking.account Property set failed: object in path "errors.validationModel.bookingSource" could not be found or was destroyed. Error: Property set failed: object in path "errors.validationModel.bookingSource" could not be found or was destroyed.
```

**Comment by Truffula** [comment]
It seems that the nesting only allows one property deep; I have it working with the equivalent of: 

```js
validations: {
    'model.name': {presence: true}
}
```
but received a similar error to @jackmatt2 with the following:
```js
validations: {
    'model.address.suburb': {presence: true}
}
```

It appears to be a limitation of `Ember.set`, in that it assumes the second-to-last string in the path (in this case `address`) is an object, and that doesn't exist.

For a workaround at the moment (with a `belongsTo` relationship) I've added this to my model:

```js
addressSuburb: Ember.computed.alias('address.suburb'),
```

and put the validation on `addressSuburb` instead.

@jackmatt2 you might also be able to alias `validationModel.bookingSource`...

**Comment by jackmatt2** [comment]
@Truffula Thanks for this.  I am unsure why I even need to alias the model in the first place.  I am only doing that because the example @danmcclain used was doing it that way.  Do you know if there is a good reason for this or can I use the model directly?

**Comment by Truffula** [comment]
@jackmatt2 If you're using a Controller rather than ObjectController (which is deprecated for performance reasons AFAIK) the controller doesn't proxy for the model, so you need to use a complete path. For single-depth validations you can refer to `model.foo` in each validation name - the alias of the model itself is unnecessary. Where you might need an alias is to access a child object without a three-part path (which is where it breaks because ember-validations creates a flat errors object, it doesn't copy the tree structure of your model):

```js
export default Ember.Controller.extend(EmberValidations.Mixin, {

  bookingSource: Ember.computed.alias("model.bookingSource"),

  validations : {
    'bookingSource.content' : { // Only two deep, should be ok
      presence: true
    },
  }
});
```

**Comment by jackmatt2** [comment]
@Truffula cheers, I have tested this now and can confirm it is working.

**Comment by drewcovi** [comment]
glad to see some other folks running into this, although I had sorted out how to get the validations caught using the validate() method, the inline-errors were failing because i had not set up any aliases.  definitely a bit of a hack. cant wait for the proper solution.

**Comment by bcardarella** [comment]
This is a top priority to have a sorted out version working by `alpha.5`. I'm split between refactoring out the observers and providing a clean ObjectController-less interface

**Comment by drewcovi** [comment]
:+1: great news Brian, thanks a billion

**Comment by Truffula** [comment]
@drewcovi I ended up going down a similar path (actually validating child objects recursively with a mixin) since my previous solution (with aliases) had limitations for deeply nested objects and hasMany properties.

Love your work, Brian!

**Comment by jamoon** [comment]
@drewcovi - I would be interested in seeing our approach for deeply nested objects.

**Comment by Truffula** [comment]
Each model has a 'children' property (an array), and a Mixin provides a custom `validate` function which calls `validate` on each child (including hasMany, belongsTo, and `ember-data-model-fragments` variants hasOneFragment and hasManyFragments), returning an `RSVP.hash`; if the children have the mixin, it happens recursively. I'm not actually interested in the specific errors it returns at the top of the recursive chain, as that's handled by the individual form components which map their own validation to the child objects themselves; the recursive function just lets me make sure everything's valid before submitting.

**Comment by kmiyashiro** [comment]
@Truffula Did you post your mixin to a gist?

**Comment by martndemus** [comment]
I am going to close this issue for now, feel free to reopen this issue if you feel like it's still relevant and we should take a look at it.
